### DISCUSSION TITLE
Make some of the other optionals required if one optional is specified

### DISCUSSION THREAD

**Discussion by vresetnikov** [colinhacks/zod#978] [Q&A]
Hello,

For example, if an optional "Address" is submitted, make other optionals such as "Postal Code" and "City" required as a result?

Could you kindly tell me what would be the best approach? 

Thank you very much

**Comment by JacobWeisenburger** [comment]
You could use either of these:
- [Unions](https://github.com/colinhacks/zod#unions)
- [Discriminated unions](https://github.com/colinhacks/zod#discriminated-unions)

Let me know if you have any questions

## union
```ts
const schema = z.union( [
    z.object( {
        street: z.string(),
        city: z.string(),
        postalCode: z.string(),
    } ).strict(), // .strict is needed for your use case to work properly
    z.object( {
        city: z.string().optional(),
        postalCode: z.string().optional(),
    } ).strict(), // .strict is needed for your use case to work properly
] )

schema.parse( {
    street: '123 Main St',
    city: 'Anytown',
    postalCode: '12345',
} ) // passes

schema.parse( {
    city: 'Anytown',
    postalCode: '12345',
} ) // passes

schema.parse( {
    street: '123 Main St',
} ) // fails
```

## discriminatedUnion
```ts
const schema = z.discriminatedUnion( 'type', [
    z.object( {
        type: z.literal( 'address' ),
        street: z.string(),
        city: z.string(),
        postalCode: z.string(),
    } ), // .strict is not needed
    z.object( {
        type: z.literal( 'contact' ),
        email: z.string().email(),
        phone: z.string().regex( /^\d{3}-\d{3}-\d{4}$/ ),
    } ), // .strict is not needed
] )

schema.parse( {
    type: 'address',
    street: '123 Main St',
    city: 'Anytown',
    postalCode: '12345',
} ) // passes

schema.parse( {
    type: 'contact',
    email: 'yay@hurray.com',
    phone: '123-456-7890',
} ) // passes

schema.parse( {
    type: 'address',
    street: '123 Main St',
} ) // fails
```

**Reply by vresetnikov** [reply]
Could you let me know if there is a solution to my final example please?

**Comment by vresetnikov** [comment]
Hello,

Thank you for your kindness in providing a response in a timely manner and sorry for me not doing the same. 

Unfortunately, I have not been able to benefit from the examples provided:

- The union example, while dealing with the original issue, had me encounter a new problem: it allows passing the 'city' and 'postalCode' when the main optional 'Address' is not provided. I understand this was not clearly expressed in my original question. Maybe my task is becoming too elaborate and should be delegated to the controller?
- I am also struggling to understand how the union works in this case, i.e. how omitting the "street" property in the second object ties it to other properties being required. 
- I was unable to try out the second example. I have imported z from zod, but the interpreter complains that property discriminatedUnion does not exist.

**Comment by JacobWeisenburger** [comment]
> The union example, while dealing with the original issue, had me encounter a new problem: it allows passing the 'city' and 'postalCode' when the main optional 'Address' is not provided. I understand this was not clearly expressed in my original question...

> I am also struggling to understand how the union works in this case, i.e. how omitting the "street" property in the second object ties it to other properties being required.

No problem. If I understand your requirements correctly, this is a really simple fix. So basically you are saying you either want an object with all of these properties or none of these properties.

```ts
const schema = z.union( [
    z.object( {
        street: z.string(),
        city: z.string(),
        postalCode: z.string(),
    } ).strict(), // .strict is needed for your use case to work properly
    z.object( {
        // this is an empty object
    } ).strict(), // .strict is needed for your use case to work properly
] )

schema.parse( {
    street: '123 Main St',
    city: 'Anytown',
    postalCode: '12345',
} ) // passes

schema.parse( { city: 'Anytown', postalCode: '12345' } ) // fails
schema.parse( { city: 'Anytown' } ) // fails
schema.parse( { postalCode: '12345' } ) // fails
schema.parse( { street: '123 Main St' } ) // fails
```

---
> I was unable to try out the second example. I have imported z from zod, but the interpreter complains that property discriminatedUnion does not exist.

`discriminatedUnion` is really new. It was released in [v3.12.0](https://github.com/colinhacks/zod/releases/tag/v3.12.0). So you just need to update your version of zod for it to work.

**Reply by vresetnikov** [reply]
Thank you for your reply! Yes, this indeed looks like it! Forgive me for disturbing you so much, it is just my first time working with Zod (or with Typescript for that matter). I am struggling to implement your solution to my schema:

```typescript
import {object, string, TypeOf, boolean, union} from "zod";
import isISODate from "../utils/dates.utils";

export const createDatesTemplateSchema = object({
    body: object({
        templateName: string({
            required_error: 'Template name is required.',
        }),
        weekends: string()
            .min(7, 'The character length should be 7.')
            .max(7, 'The character length should be 7.')
            .optional(),
        from: string().refine(isISODate,{message:'Invalid ISO string provided.'}).optional(),
        to: string().refine(isISODate,{message:'Invalid ISO string provided.'}).optional(),
        description: string().optional(),
        isActive: boolean().optional(),
        dates: object(
            {
                date: string().refine(isISODate,{message:'Invalid ISO string provided.'}),
                active: boolean(),
                description: string(),
            }
        ).array().optional(),
    }).refine(
        data => data.weekends || data.dates,
        'Either the weekends or the dates must be supplied.'
    )
});
```

Here, **weekends** is optional and the optionals _**from**_, _**to**_, _**description**_ and _**isActive**_ would be required/ accepted only if **weekends** is supplied. 

Could you kindly tell me what would be the correct way to insert your proposed solution while keeping all the other properties unaffected?

**Reply by vresetnikov** [reply]
Today I have tried making the union separate and then merging it with my main schema object, but typescript won't let me. I have looked at the documentation, but I cannot find anything that would match my case.
